File: app/src/main/java/com/example/alarm/SettingsFragment.java
package com.example.alarm;

import android.annotation.SuppressLint;
import android.media.MediaPlayer;
import android.os.Bundle;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.PopupMenu;
import android.widget.PopupWindow;

import androidx.core.content.ContextCompat;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.alarm.databinding.FragmentSettingsBinding;
import com.example.alarm.model.Settings;

import java.util.Arrays;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import dagger.hilt.android.AndroidEntryPoint;


@AndroidEntryPoint
public class SettingsFragment extends Fragment {

    private FragmentSettingsBinding binding;
    private long globalId;
    private MediaPlayer mediaPlayer;
    private AlarmViewModel alarmViewModel;
    private final ExecutorService executorService = Executors.newSingleThreadExecutor();

    @SuppressLint("DiscouragedApi")
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        alarmViewModel = new ViewModelProvider(this).get(AlarmViewModel.class);
        binding = FragmentSettingsBinding.inflate(inflater, container, false);
        alarmViewModel.registerPreferences(requireContext());

        // Update wallpapers and preferences
        executorService.submit(() -> {
            Settings settings = alarmViewModel.getSettings();
            requireActivity().runOnUiThread(() -> {
                binding.melodyName.setText(settings.getMelody());
                binding.repeatRadioGroup.setEnabled(settings.getRepetitions() == 1);
                binding.switchVibration.setChecked(settings.getVibration());
                switch (settings.getRepetitions()) {
                    case 3:
                        binding.repeats3.setChecked(true);
                        break;
                    case 5:
                        binding.repeats5.setChecked(true);
                        break;
                    default:
                        binding.repeatsInfinite.setChecked(true);
                        break;
                }
                switch (settings.getInterval()) {
                    case 3:
                        binding.interval3.setChecked(true);
                        break;
                    case 5:
                        binding.interval5.setChecked(true);
                        break;
                    default:
                        binding.interval10.setChecked(true);
                        break;
                }
                globalId = settings.getId();
            });
        });

        binding.changeMelody.setOnClickListener(this::showSignalsPopupMenu);
        binding.playMelody.setOnClickListener(v -> executorService.submit(() -> {
            String melody = alarmViewModel.getSettings().getMelody();
            playMelody(getMelodyResource(melody));
        }));

        binding.switchVibration.setOnClickListener(v -> {
            Settings s = readSettings(globalId);
            executorService.submit(() -> alarmViewModel.updateSettings(s));
        });

        setUpRepeatSettings();
        setUpIntervalSettings();

        binding.changeColorTheme.setOnClickListener(v -> showColorThemePopupMenu(v, container));
        binding.changeWallpaper.setOnClickListener(v -> showWallpapersPopupMenu(v, container));

        return binding.getRoot();
    }

    private void setUpRepeatSettings() {
        binding.repeats3.setOnClickListener(v -> updateSettings());
        binding.repeats5.setOnClickListener(v -> updateSettings());
        binding.repeatsInfinite.setOnClickListener(v -> updateSettings());
    }

    private void setUpIntervalSettings() {
        binding.interval3.setOnClickListener(v -> {
            updateSettings();
            updatePrefs(3);
        });
        binding.interval5.setOnClickListener(v -> {
            updateSettings();
            updatePrefs(5);
        });
        binding.interval10.setOnClickListener(v -> {
            updateSettings();
            updatePrefs(10);
        });
    }

    private void updateSettings() {
        Settings s = readSettings(globalId);
        executorService.submit(() -> alarmViewModel.updateSettings(s));
    }

    private Settings readSettings(long id) {
        Settings settings = new Settings(id);
        settings.setMelody(binding.melodyName.getText().toString());
        settings.setVibration(binding.switchVibration.isChecked());
        settings.setInterval(binding.interval3.isChecked() ? 3 :
                binding.interval5.isChecked() ? 5 : 10);
        settings.setRepetitions(binding.repeats3.isChecked() ? 3 :
                binding.repeats5.isChecked() ? 5 : 100);
        return settings;
    }

    private void updatePrefs(int interval) {
        alarmViewModel.editPreferencesInterval(requireContext(), interval);
    }

    private String getMelodyName(int itemId) {
        if (itemId == R.id.melody1) return getString(R.string.melody1);
        else if (itemId == R.id.melody2) return getString(R.string.melody2);
        else if (itemId == R.id.melody3) return getString(R.string.melody3);
        else if (itemId == R.id.melody4) return getString(R.string.melody4);
        else if (itemId == R.id.melody5) return getString(R.string.melody5);
        else if (itemId == R.id.melody6) return getString(R.string.melody6);
        else if (itemId == R.id.melody7) return getString(R.string.melody7);
        else if (itemId == R.id.melody8) return getString(R.string.melody8);
        else if (itemId == R.id.melody9) return getString(R.string.melody9);
        else if (itemId == R.id.melody10) return getString(R.string.melody10);
        else if (itemId == R.id.melody11) return getString(R.string.melody11);
        else if (itemId == R.id.melody12) return getString(R.string.melody12);
        else return null;
    }

    private int getMelodyResource(String melody) {
        switch (melody) {
            case "melody1": return R.raw.default_signal1;
            case "melody2": return R.raw.default_signal2;
            case "melody3": return R.raw.default_signal3;
            case "melody4": return R.raw.default_signal4;
            case "melody5": return R.raw.default_signal5;
            case "melody6": return R.raw.signal;
            case "melody7": return R.raw.banjo_signal;
            case "melody8": return R.raw.morning_signal;
            case "melody9": return R.raw.simple_signal;
            case "melody10": return R.raw.fitness_signal;
            case "melody11": return R.raw.medieval_signal;
            case "melody12": return R.raw.introduction_signal;
            default: return -1; // Melody not selected
        }
    }

    private void showWallpapersPopupMenu(View view, ViewGroup container) {
        // Inflate the custom layout for the popup
        LayoutInflater inflater = LayoutInflater.from(requireContext());
        View popupView = inflater.inflate(R.layout.popup_menu_wallpaper_layout, container, false);

        // Create the PopupWindow
        PopupWindow popupWindow = new PopupWindow(
                popupView,
                LinearLayout.LayoutParams.WRAP_CONTENT,
                LinearLayout.LayoutParams.WRAP_CONTENT,
                true // Focusable
        );

        // Show the popup window at a specific location
        popupWindow.showAtLocation(requireView(), Gravity.CENTER, 0, 0);

        // Set up the RecyclerView
        RecyclerView recyclerView = popupView.findViewById(R.id.recycler_view);
        recyclerView.setLayoutManager(new LinearLayoutManager(requireContext()));

        // Define the menu items
        List<MenuItemData> menuItems = Arrays.asList(
                new MenuItemData(getString(R.string.classic), R.drawable.whitequad),
                new MenuItemData("1.", R.drawable.wallpaper1),
                new MenuItemData("2.", R.drawable.wallpaper2),
                new MenuItemData("3.", R.drawable.wallpaper3),
                new MenuItemData("4.", R.drawable.wallpaper4),
                new MenuItemData("5.", R.drawable.wallpaper5),
                new MenuItemData("6.", R.drawable.wallpaper6),
                new MenuItemData("7.", R.drawable.wallpaper7),
                new MenuItemData("8.", R.drawable.wallpaper8),
                new MenuItemData("9.", R.drawable.wallpaper9),
                new MenuItemData("10.", R.drawable.wallpaper10)
        );

        // Create and set the adapter
        PopupMenuWallpaperAdapter adapter = new PopupMenuWallpaperAdapter(menuItems, menuItem -> {
            String selectedWallpaper = "";
            switch (menuItem.title) {
                case "1.":
                    selectedWallpaper = getString(R.string.wallpaper1);
                    break;
                case "2.":
                    selectedWallpaper = getString(R.string.wallpaper2);
                    break;
                case "3.":
                    selectedWallpaper = getString(R.string.wallpaper3);
                    break;
                case "4.":
                    selectedWallpaper = getString(R.string.wallpaper4);
                    break;
                case "5.":
                    selectedWallpaper = getString(R.string.wallpaper5);
                    break;
                case "6.":
                    selectedWallpaper = getString(R.string.wallpaper6);
                    break;
                case "7.":
                    selectedWallpaper = getString(R.string.wallpaper7);
                    break;
                case "8.":
                    selectedWallpaper = getString(R.string.wallpaper8);
                    break;
                case "9.":
                    selectedWallpaper = getString(R.string.wallpaper9);
                    break;
                case "10.":
                    selectedWallpaper = getString(R.string.wallpaper10);
                    break;
                case "Classic":
                    selectedWallpaper = ""; // Default case
                    break;
            }
            // Save the selected wallpaper
            alarmViewModel.editPreferencesWallpaper(requireContext(), selectedWallpaper);
            // Dismiss the popup
            popupWindow.dismiss();
        });

        recyclerView.setAdapter(adapter);

        // Show the popup below the clicked view
        popupWindow.showAsDropDown(view);
    }


    private void showColorThemePopupMenu(View view, ViewGroup container) {
        LayoutInflater inflater = LayoutInflater.from(requireContext());
        View popupView = inflater.inflate(R.layout.popup_menu_wallpaper_layout, container, false);

        PopupWindow popupWindow = new PopupWindow(
                popupView,
                LinearLayout.LayoutParams.WRAP_CONTENT,
                LinearLayout.LayoutParams.WRAP_CONTENT,
                true
        );

        popupWindow.showAtLocation(requireView(), Gravity.CENTER, 0, 0);

        RecyclerView recyclerView = popupView.findViewById(R.id.recycler_view);
        recyclerView.setLayoutManager(new LinearLayoutManager(requireContext()));

        List<ColorThemeMenuItem> menuItems = Arrays.asList(
                new ColorThemeMenuItem(R.color.colorPrimary, R.color.colorAccent, 0),
                new ColorThemeMenuItem(R.color.color1_main, R.color.color1_secondary, 1),
                new ColorThemeMenuItem(R.color.color2_main, R.color.color2_secondary, 2),
                new ColorThemeMenuItem(R.color.color3_main, R.color.color3_secondary, 3),
                new ColorThemeMenuItem(R.color.color4_main, R.color.color4_secondary, 4),
                new ColorThemeMenuItem(R.color.color5_main, R.color.color5_secondary, 5),
                new ColorThemeMenuItem(R.color.color6_main, R.color.color6_secondary, 6),
                new ColorThemeMenuItem(R.color.color7_main, R.color.color7_secondary, 7),
                new ColorThemeMenuItem(R.color.color8_main, R.color.color8_secondary, 8)
        );

        ColorThemeMenuAdapter adapter = new ColorThemeMenuAdapter(menuItems, menuItem -> {
            alarmViewModel.editPreferencesTheme(requireContext(), menuItem.getThemeNumber());
            requireActivity().recreate();
            popupWindow.dismiss();
        });

        recyclerView.setAdapter(adapter);

        popupWindow.showAsDropDown(view);
    }


    private void onWallpaperItemSelected(MenuItemData menuItem) {
        String temp = menuItem.title.equals("Classic") ? "" : "wallpaper" + menuItem.title.charAt(0);
        alarmViewModel.editPreferencesWallpaper(requireContext(), temp);
    }

    @SuppressLint("DiscouragedApi")
    private void updateWallpapers(String wallpaper) {
        binding.wallpaperName.setText(wallpaper);
        if (!wallpaper.isEmpty()) {
            int resId = requireContext().getResources().getIdentifier(wallpaper, "drawable", requireContext().getPackageName());
            if (resId != 0) {
                binding.settingsLayout.setBackground(ContextCompat.getDrawable(requireContext(), resId));
            }
        } else {
            binding.settingsLayout.setBackground(null);
            binding.wallpaperName.setText("Classic");
        }
    }

    private void playMelody(int signalId) {
        if (mediaPlayer != null && mediaPlayer.isPlaying()) {
            stopAndReleaseMediaPlayer();
        }

        mediaPlayer = MediaPlayer.create(getContext(), signalId);
        mediaPlayer.start();
        mediaPlayer.setOnCompletionListener(mp -> {
            stopAndReleaseMediaPlayer();
            binding.floatingActionButtonVolumeOff.setVisibility(View.GONE);
        });
        binding.floatingActionButtonVolumeOff.setVisibility(View.VISIBLE);
        binding.floatingActionButtonVolumeOff.setOnClickListener(v -> stopAndReleaseMediaPlayer());
    }

    private void stopAndReleaseMediaPlayer() {
        if (mediaPlayer != null) {
            if (mediaPlayer.isPlaying()) {
                mediaPlayer.stop();
            }
            mediaPlayer.release();
            mediaPlayer = null;
        }
    }
    private void showSignalsPopupMenu(View view) {
        PopupMenu popupMenu = new PopupMenu(requireContext(), view);
        popupMenu.getMenuInflater().inflate(R.menu.melody_menu, popupMenu.getMenu());
        popupMenu.setOnMenuItemClickListener(menuItem -> {
            String newMelodyName = getMelodyName(menuItem.getItemId());
            if (newMelodyName != null) {
                binding.melodyName.setText(newMelodyName);
                Settings settings = readSettings(globalId);
                executorService.submit(() -> alarmViewModel.updateSettings(settings));
                return true;
            }
            return false;
        });
        popupMenu.show();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        alarmViewModel.unregisterPreferences(requireContext());
        executorService.shutdown();
    }

}

File: app/src/main/java/com/example/alarm/SettingsFragment.kt
package com.example.alarm

import android.annotation.SuppressLint
import android.content.Context
import android.content.Context.LAYOUT_INFLATER_SERVICE
import android.content.SharedPreferences
import android.media.MediaPlayer
import android.os.Bundle
import android.preference.PreferenceManager
import android.util.Log
import android.view.Gravity
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.PopupMenu
import android.widget.PopupWindow
import android.widget.Toast
import androidx.core.content.ContextCompat
import androidx.core.content.ContextCompat.getSystemService
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.alarm.databinding.FragmentSettingsBinding
import com.example.alarm.model.AlarmService
import com.example.alarm.model.Settings
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.async
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class SettingsFragment : Fragment() {

    private lateinit var binding: FragmentSettingsBinding
    private var globalId: Long = 0
    private lateinit var mediaPlayer: MediaPlayer
    private lateinit var alarmViewModel: AlarmViewModel

    @SuppressLint("DiscouragedApi")
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        binding = FragmentSettingsBinding.inflate(inflater, container, false)
        alarmViewModel = ViewModelProvider(requireActivity())[AlarmViewModel::class.java]
        alarmViewModel.registerPreferences(requireContext())
        alarmViewModel.wallpaper.observe(viewLifecycleOwner) {
            updateWallpapers(it)
        }
        val (wallpaper, _) = alarmViewModel.getPreferencesWallpaperAndInterval(requireContext())
        if(wallpaper != "") {
            binding.wallpaperName.text = wallpaper
            val resId = resources.getIdentifier(wallpaper, "drawable", requireContext().packageName)
            if(resId != 0) binding.settingsLayout.background = ContextCompat.getDrawable(requireContext(), resId)
        }
        else {
            binding.wallpaperName.text = "Classic"
        }
        val themeNumber = alarmViewModel.getPreferencesTheme(requireContext())
        if(themeNumber != 0) binding.colorThemeName.text = "Theme ${themeNumber+1}" else binding.colorThemeName.text = "Classic"
        lifecycleScope.launch {
            val settings = alarmViewModel.getSettings()
            binding.melodyName.text = settings.melody
            binding.repeatRadioGroup.isEnabled = settings.repetitions == 1
            binding.switchVibration.isChecked = settings.vibration == 1
            when(settings.repetitions) {
                3 -> binding.repeats3.isChecked = true
                5 -> binding.repeats5.isChecked = true
                else -> binding.repeatsInfinite.isChecked = true
            }
            when(settings.interval) {
                3 -> binding.interval3.isChecked = true
                5 -> binding.interval5.isChecked = true
                else -> binding.interval10.isChecked = true
            }
            globalId = settings.id
        }
        binding.changeMelody.setOnClickListener {
            showSignalsPopupMenu(it)
        }
        binding.playMelody.setOnClickListener {
            lifecycleScope.launch {
                val settings = async(Dispatchers.IO) { alarmViewModel.getSettings() }
                when (settings.await().melody) {
                    getString(R.string.melody1) -> playMelody(R.raw.default_signal1)
                    getString(R.string.melody2) -> playMelody(R.raw.default_signal2)
                    getString(R.string.melody3) -> playMelody(R.raw.default_signal3)
                    getString(R.string.melody4) -> playMelody(R.raw.default_signal4)
                    getString(R.string.melody5) -> playMelody(R.raw.default_signal5)
                    getString(R.string.melody6) -> playMelody(R.raw.signal)
                    getString(R.string.melody7) -> playMelody(R.raw.banjo_signal)
                    getString(R.string.melody8) -> playMelody(R.raw.morning_signal)
                    getString(R.string.melody9) -> playMelody(R.raw.simple_signal)
                    getString(R.string.melody10) -> playMelody(R.raw.fitness_signal)
                    getString(R.string.melody11) -> playMelody(R.raw.medieval_signal)
                    getString(R.string.melody12) -> playMelody(R.raw.introduction_signal)
                    else -> {
                        Toast.makeText(requireContext(), "Melody not selected", Toast.LENGTH_SHORT).show()
                    }
                }
            }
        }
        binding.switchVibration.setOnClickListener {
            val s = readSettings(globalId)
            lifecycleScope.launch { alarmViewModel.updateSettings(s) }
        }
        binding.repeats3.setOnClickListener {
            val s = readSettings(globalId)
            lifecycleScope.launch { alarmViewModel.updateSettings(s) }
        }
        binding.repeats5.setOnClickListener {
            val s = readSettings(globalId)
            lifecycleScope.launch { alarmViewModel.updateSettings(s) }
        }
        binding.repeatsInfinite.setOnClickListener {
            val s = readSettings(globalId)
            lifecycleScope.launch { alarmViewModel.updateSettings(s) }
        }
        binding.interval3.setOnClickListener {
            val s = readSettings(globalId)
            lifecycleScope.launch { alarmViewModel.updateSettings(s) }
            updatePrefs(3)
        }
        binding.interval5.setOnClickListener {
            val s = readSettings(globalId)
            lifecycleScope.launch { alarmViewModel.updateSettings(s) }
            updatePrefs(5)
        }
        binding.interval10.setOnClickListener {
            val s = readSettings(globalId)
            lifecycleScope.launch { alarmViewModel.updateSettings(s) }
            updatePrefs(10)
        }
        binding.changeColorTheme.setOnClickListener {
            showColorThemePopupMenu(it, container as ViewGroup)
        }
        binding.changeWallpaper.setOnClickListener {
            showWallpapersPopupMenu(it, container as ViewGroup)
        }
        return binding.root
    }

    private fun readSettings(id: Long): Settings {
        val settings = Settings(
            id = id,
            melody = binding.melodyName.text.toString(),
            vibration = if(binding.switchVibration.isChecked) 1 else 0,
            interval = if(binding.interval3.isChecked) 3
            else if(binding.interval5.isChecked) 5
            else 10,
            repetitions = if(binding.repeats3.isChecked) 3
            else if(binding.repeats5.isChecked) 5
            else 100,
            disableType = 0 //todo
        )
        return settings
    }

    private fun updatePrefs(interval: Int) {
        alarmViewModel.editPreferencesInterval(requireContext(), interval)
    }
    private fun showSignalsPopupMenu(view: View) {
        val popupMenu = PopupMenu(requireContext(), view)
        popupMenu.menuInflater.inflate(R.menu.melody_menu, popupMenu.menu)
        popupMenu.setOnMenuItemClickListener { menuItem: MenuItem ->
            val newMelodyName = when (menuItem.itemId) {
                R.id.melody1 -> getString(R.string.melody1)
                R.id.melody2 -> getString(R.string.melody2)
                R.id.melody3 -> getString(R.string.melody3)
                R.id.melody4 -> getString(R.string.melody4)
                R.id.melody5 -> getString(R.string.melody5)
                R.id.melody6 -> getString(R.string.melody6)
                R.id.melody7 -> getString(R.string.melody7)
                R.id.melody8 -> getString(R.string.melody8)
                R.id.melody9 -> getString(R.string.melody9)
                R.id.melody10 -> getString(R.string.melody10)
                R.id.melody11 -> getString(R.string.melody11)
                R.id.melody12 -> getString(R.string.melody12)
                else -> null
            }

            newMelodyName?.let { name ->
                binding.melodyName.text = name
                val s = readSettings(globalId)
                lifecycleScope.launch { alarmViewModel.updateSettings(s) }
                true
            } ?: false
        }
        popupMenu.show()
    }
    private fun showWallpapersPopupMenu(view: View, container: ViewGroup) {
        val inflater = layoutInflater
        val popupView = inflater.inflate(R.layout.popup_menu_wallpaper_layout, container, false)

        val popupWindow = PopupWindow(
            popupView,
            LinearLayout.LayoutParams.WRAP_CONTENT,
            LinearLayout.LayoutParams.WRAP_CONTENT,
            true
        )

        popupWindow.showAtLocation(requireView(), Gravity.CENTER, 0, 0)
        val recyclerView = popupView.findViewById<RecyclerView>(R.id.recycler_view)
        recyclerView.layoutManager = LinearLayoutManager(requireContext())

        val menuItems = listOf(
            MenuItemData("Classic", R.drawable.whitequad),
            MenuItemData("1.", R.drawable.wallpaper1),
            MenuItemData("2.", R.drawable.wallpaper2),
            MenuItemData("3.", R.drawable.wallpaper3),
            MenuItemData("4.", R.drawable.wallpaper4),
            MenuItemData("5.", R.drawable.wallpaper5),
            MenuItemData("6.", R.drawable.wallpaper6),
            MenuItemData("7.", R.drawable.wallpaper7),
            MenuItemData("8.", R.drawable.wallpaper8),
            MenuItemData("9.", R.drawable.wallpaper9),
            MenuItemData("10.", R.drawable.wallpaper10)
        )
        var temp = ""
        val adapter = PopupMenuWallpaperAdapter(menuItems) { menuItem ->
            temp = when(menuItem.title) {
                "Classic" -> ""
                "1." -> "wallpaper1"
                "2." -> "wallpaper2"
                "3." -> "wallpaper3"
                "4." -> "wallpaper4"
                "5." -> "wallpaper5"
                "6." -> "wallpaper6"
                "7." -> "wallpaper7"
                "8." -> "wallpaper8"
                "9." -> "wallpaper9"
                "10." -> "wallpaper10"
                else -> ""
            }
            alarmViewModel.editPreferencesWallpaper(requireContext(), temp)
            popupWindow.dismiss()
        }

        recyclerView.adapter = adapter

        popupWindow.showAsDropDown(view)
    }
    private fun showColorThemePopupMenu(view: View, container: ViewGroup) {
        val inflater = layoutInflater
        val popupView = inflater.inflate(R.layout.popup_menu_wallpaper_layout, container, false)

        val popupWindow = PopupWindow(
            popupView,
            LinearLayout.LayoutParams.WRAP_CONTENT,
            LinearLayout.LayoutParams.WRAP_CONTENT,
            true
        )

        popupWindow.showAtLocation(requireView(), Gravity.CENTER, 0, 0)
        val recyclerView = popupView.findViewById<RecyclerView>(R.id.recycler_view)
        recyclerView.layoutManager = LinearLayoutManager(requireContext())

        val menuItems = listOf(
            ColorThemeMenuItem(R.color.colorPrimary, R.color.colorAccent, 0),
            ColorThemeMenuItem(R.color.color1_main, R.color.color1_secondary, 1),
            ColorThemeMenuItem(R.color.color2_main, R.color.color2_secondary, 2),
            ColorThemeMenuItem(R.color.color3_main, R.color.color3_secondary, 3),
            ColorThemeMenuItem(R.color.color4_main, R.color.color4_secondary, 4),
            ColorThemeMenuItem(R.color.color5_main, R.color.color5_secondary, 5),
            ColorThemeMenuItem(R.color.color6_main, R.color.color6_secondary, 6),
            ColorThemeMenuItem(R.color.color7_main, R.color.color7_secondary, 7),
            ColorThemeMenuItem(R.color.color8_main, R.color.color8_secondary, 8)
        )
        val adapter = ColorThemeMenuAdapter(menuItems) { menuItem ->
            alarmViewModel.editPreferencesTheme(requireContext(), menuItem.themeNumber)
            requireActivity().recreate()
            popupWindow.dismiss()
        }

        recyclerView.adapter = adapter

        popupWindow.showAsDropDown(view)
    }

    @SuppressLint("DiscouragedApi")
    private fun updateWallpapers(wallpaper: String) {
        binding.wallpaperName.text = wallpaper
        if(wallpaper != "") {
            val resId = resources.getIdentifier(wallpaper, "drawable", requireContext().packageName)
            if(resId != 0)
                binding.settingsLayout.background = ContextCompat.getDrawable(requireContext(), resId)
        }
        else {
            binding.settingsLayout.background = null
            binding.wallpaperName.text = "Classic"
        }
    }

    private fun playMelody(signalId: Int) {
        if (::mediaPlayer.isInitialized) {
            stopAndReleaseMediaPlayer()
        }

        mediaPlayer = MediaPlayer.create(context, signalId)
        mediaPlayer.start()
        mediaPlayer.setOnCompletionListener {
            stopAndReleaseMediaPlayer()
            binding.floatingActionButtonVolumeOff.visibility = View.GONE
        }
        binding.floatingActionButtonVolumeOff.visibility = View.VISIBLE
        binding.floatingActionButtonVolumeOff.setOnClickListener {
            stopAndReleaseMediaPlayer()
            binding.floatingActionButtonVolumeOff.visibility = View.GONE
        }
    }

    private fun stopAndReleaseMediaPlayer() {
        if (::mediaPlayer.isInitialized) {
            if (mediaPlayer.isPlaying) {
                mediaPlayer.stop()
            }
            mediaPlayer.release()
            mediaPlayer = MediaPlayer()
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        alarmViewModel.unregisterPreferences(requireContext())
    }
}